spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/DataFlowOps.scala

```diff
@@ -0,0 +1,586 @@
+ * this splits values into a TensorArray with T tensors.
+ * TensorArray index t will be the subtensor of values with starting position
+ *   `(n0 + n1 + ... + n(t-1), 0, 0, ...)```
+ * and having size
+ *   nt x d0 x d1 x ...
+ *
+ * @tparam T Model parameter numeric type.
+ * @tparam D Element numeric type in the tensor array.
+ */
+private[bigdl] class TensorArraySplit[T: ClassTag, D: ClassTag]()(
+  implicit ev: TensorNumeric[T], ev2: TensorNumeric[D])
+  extends Operation[Table, Tensor[Float], T] {
+
+  output = TensorArray.FlowOut
+
+  override def updateOutput(input: Table): Tensor[Float] = {
+    val handle = input[Tensor[String]](1)
+    val value = input[Tensor[D]](2)
+    val lengths = input[Tensor[Int]](3)
+    require(handle.isScalar, "Handle of a TensorArray must be a scalar")
+    require(lengths.nDimension() == 1, "lengths must be a vector")
+
+    val tensorArray = TensorArray[D](handle.value())
+
+    var i = 1
+    var index = 1
+    while(i <= lengths.size(1)) {
+      tensorArray(i - 1) = value.narrow(1, index, lengths.valueAt(i))
+      index += lengths.valueAt(i)
+      i += 1
+    }
+
+    output
+  }
+}
+
+/**
+ * Get the current size of the TensorArray.
+ *
+ * @tparam T Model parameter numeric type.
+ */
+private[bigdl] class TensorArraySize[T: ClassTag]()(implicit ev: TensorNumeric[T])
+  extends Operation[Table, Tensor[Int], T] {
+
+  override def updateOutput(input: Table): Tensor[Int] = {
+    val handle = input[Tensor[String]](1)
+    require(handle.isScalar, "Handle of a TensorArray must be a scalar")
+
+    val tensorArray = TensorArray(handle.value())
+
+    output = Tensor.scalar[Int](tensorArray.size)
+    output
+  }
+}
+
+/**
+ * Delete the TensorArray from the context.
+ *
+ * @tparam T Model parameter numeric type.
+ */
+private[bigdl] class TensorArrayClose[T: ClassTag]()(implicit ev: TensorNumeric[T])
+  extends Operation[Tensor[String], Tensor[Float], T] {
+
+  output = TensorArray.FlowOut
+
+  override def updateOutput(input: Tensor[String]): Tensor[Float] = {
+    require(input.isScalar, "Handle of a TensorArray must be a scalar")
+    TensorArray.release(input.value())
+    output
+  }
+}
+
+private[bigdl] class Stack[D](maxSize: Int) {
+  private var count = 0
+  private val tensors = new ArrayBuffer[Tensor[D]]()
+
+  def pop(): Tensor[D] = {
+    require(count > 0, "There's no tensors in the stack")
+    count -= 1
+    tensors.remove(count)
+  }
+
+  def push(t: Tensor[D]): Unit = {
+    require(count < maxSize, "Stack is full")
+    tensors.append(t.clone())
+    count += 1
+  }
+}
+
+private[bigdl] object Stack {
+  private val stacks = new ConcurrentHashMap[String, Stack[_]]()
+
+  def apply[D](key: String): Stack[D] = {
+    require(stacks.containsKey(key), s"Cannot find Stack for name $key")
+    stacks.get(key).asInstanceOf[Stack[D]]
+  }
+
+  def update(key: String, value: Stack[_]): Unit = {
+    stacks.put(key, value)
+  }
+
+  def release(key : String): Unit = {
```

Comment:
This method is not called anywhere, maybe a leak?

Reply:
fixed